### PR TITLE
Add Package Level Documentation to o.e.r.blobstore

### PR DESCRIPTION
**Pull request by original-brownbear**
* Added verbose documentation for the `o.e.r.blobstore` package similar to
that added for the snapshot package in
https://github.com/elastic/elasticsearch/pull/38108
* Moved the documentation on the BlobStoreRepository to the package
level to have things in a single place for easier readability.

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
Thanks Armin, I left a few small suggestions.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Actually neither is required for a read-only repo :) You only need GET for those, will adjust accordingly :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: I just dropped the part in brackets here now, it seemed really redundant? Given that the text below eventually just links to the classes that get serialized here anyway?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: fair point :) fixed

**Comment by original-brownbear** [comment]
thanks for taking a look @DaveCTurner, all points addressed now I think :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: What about DELETE?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: For some reason the link to snapshots package does not work, is it merged already?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Do we also have a class name that we can reference here?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: ... which is different from index id in Elasticseach cluster

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Can we elaborate a bit here why we prefer listing to reading index.latest blob?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (andrershov) [inline server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Can we elaborate here why we need both BlobStoreIndexShardSnapshot and BlobStoreIndexShardSnapshots? AFAIK, it's needed exclusively for delete snapshot operations.

**Review by ywelsch — APPROVED** [review]
@andrershov should have a look here as well before merging

**Review by andrershov — CHANGES_REQUESTED** [review]
@original-brownbear thanks for your PR, this documentation is really useful! I've asked for some clarifications/elaborations.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: It works for me in the rendered HTML version. It doesn't work in IDEA though, it just opens the package in the file tree, but the HTML is fine I think.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Unfortunately not. This is some weird list of `SnapshotId` that is read into an artificial `RepositoryData`.
See https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/repositories/RepositoryData.java#L461 
Since this isn't so relevant to the operation of the blobstore (if at all relevant at this point for that matter), I'd just skip documenting this in more detail preferably.

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Right fair point :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Phew :) I'm not sure how to answer that. You could make the argument that on S3 updating a blob is even less consistent than listing potentially but it's not really a good argument tbh. I would just chalk this up to this working the way it does for historical reasons.
As discussed last week, we will introduce a more stable way of tracking the latest `N` here via the cluster state soon and can then adjust docs?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: We currently use `BlobStoreIndexShardSnapshots` to determine the segments to upload when creating incremental snapshots of a shard as well as what to delete.
Technically speaking `BlobStoreIndexShardSnapshots` is redundant as you could rebuild it from all the `BlobStoreIndexShardSnapshot` blobs. In practice this might get tricky since it implies downloading potentially hundreds of them (you have exactly as many as you have snapshots referencing the shard).
I think I'd really just leave the details of what is in `BlobStoreIndexShardSnapshot` precisely to the link to that class (it contains some additional metadata about the time it took to snapshot the shard that we expose via an API). It seems to me it's easier to read this by not mentioning the duplication of data here, that just has the reader wondering why (and it's a fair question :D) when understanding that the `BlobStoreIndexShardSnapshots` holds the list of segments is all that really matters for the big picture here?

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: added comment in 
ad3f147

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: fixed in 
ad3f147

**Comment by original-brownbear** [comment]
@andrershov all addressed I think :)

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Agree, since we don't have a good explanation, let's skip it for now

File: server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java
Comment (andrershov) [reply server/src/main/java/org/elasticsearch/repositories/blobstore/package-info.java]: Ok

**Review by andrershov — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Jenkins run 
elasticsearch-ci/1

**Comment by original-brownbear** [comment]
thanks @andrershov @ywelsch @DaveCTurner !

**Pull request merged**
